exSpy Split

CSSFrancis opened the pull request:
### Description of the change
This is my (best) attempt at splitting the EDS and EELS. 

### Progress of the PR
- [x] Move Signals to exspy
- [x] Move Models to exspy
- [x] Move Components to exspy
- [x] Move EdgesRange to exspy (Wait till Marker Collection is Merged?)
- [x] Move Utils to espy
- [x] Move EELS/EDS Drawing tests to Spectrospy (Wait till Marker Collection is Merged?)
- [x] Move EDS/EELS Tests
- [x] Move model convolution to exSpy
- [x] Skip all Tests in hyperspy if exspy isn't installed and required.
- [x] Get Tests back to passing 
- [x] Move Documentation
- [x] Rename to exSpy


Current State:

Any PRS are welcome here.  Most of the splitting/ testing is done but there is quite a bit in the hyperspy tests that are still dependent on spectrospy.  I don't think that is a huge deal as we can always split, require spectrospy as a testing requirement and then clean up those tests later.

CSSFrancis committed:
Setup: Initial Folders Created

CSSFrancis committed:
Refactor: Moved EELS Components

CSSFrancis committed:
Moved: Moved EDS and EELS Signals to spectrospy

CSSFrancis committed:
Moved: Moved EDS and EELS Models to spectrospy

CSSFrancis committed:
Moved: Moved EELS tools to spectrospy

CSSFrancis committed:
Moved: EDS misc to spectrospy

CSSFrancis committed:
Moved: Moved signal tests

CSSFrancis committed:
Moved: Moved Model and component testing

CSSFrancis committed:
Moved: Moved Rest of dielectic function and eds utils to spectrospy

CSSFrancis committed:
Refactor: changed hyperspy references to spectrospy

CSSFrancis committed:
Refactor: Moved Tests

CSSFrancis committed:
Refactor: Refactored Tests

CSSFrancis committed:
Refactor: Fixed hyperspy_extension.yaml for spectrospy

CSSFrancis committed:
Refactor: Addded README.rst

CSSFrancis committed:
Tests: Cleaned up spectrospy tests to passing

CSSFrancis committed:
Setup: Removed EDS/EELS from setup.py

CSSFrancis committed:
Setup: Removed EDS/EELS toolkeys from extension.yaml

CSSFrancis commented:
This appears to be failing due to the GUI toolkit.  I'll need to look into this a little bit more...

CSSFrancis committed:
GUI: Moved GUI Toolkits

CSSFrancis committed:
Setup: Moved gui functions to spectrospy

CSSFrancis committed:
Testing: Trying not editable

CSSFrancis commented:
@ericpre for some reason the spectrospy package isn't getting loaded properly. I think it is because of the GUI components but I have the annoying problem of it not failing on my computer... 

Do you have any ideas about how to fix this?

CSSFrancis commented:
My current plan is that we should be able to `git-filter-repo` the commits from `hyperspy`,` hyperspy_gui_traitsui` and `hyperspy_gui_ipywidgets` so that all of the EELS EDS stuff is in one place.

CSSFrancis committed:
Testing: Added pip install to azure-pipelines.yml

CSSFrancis committed:
Documentation: Moved some documetation files and added .spectropy to build docs

CSSFrancis committed:
Documentation: Cleaning up some documentation messages

CSSFrancis committed:
Documentation: Moved images and some files

CSSFrancis committed:
Documentation: Moved linear fitting and changed docstrings pointer

CSSFrancis committed:
Copyright :Added spectrospy copyright

CSSFrancis committed:
Testing: Fixed linear model testing

CSSFrancis committed:
Moved model fitting example

francisco-dlp commented:
Great job!

Yesterday @ericpre and I discussed offline the name of the splitted package. We thought that it would be good that the name incorporates the idea that it is an EM package, basically devoted to EELS and EDX. So, @ericpre suggested adding an `e` (electron, EELS) (`EspectroSpy`), and I suggested replacing the first `s` with `x` (EDX) `EXpectroSpy`. What do you think?

CSSFrancis commented:
@francisco-dlp I don't have terribly strong opinions on any name really. I'm not really a huge fan of spectrospy cause it is very close to spectroscopy, I just needed a place holder for now.

Something shorter might be nice even if it's not the most accurate.

- Exspy
- Espy
- Spexspy

Expectrospy sounds like a spell from Harry Potter :) 

My vote would be for Expectrospy or just exspy or espy.

I like the exspy maybe the best because it also gives the idea that it used to be part of hyperspy.

ericpre commented:
> Expectrospy sounds like a spell from Harry Potter :)

Agree, and it is rather on the long side!

`EXSpy` sounds good to me and it is easy to read/pronounce (EX-Spy)?
An alternative may be `EelXSpy` (pronounciation: Eel-X-Spy; il-ɛks-spaɪ) or `EleXSpy` (EleX-Spy; ɛlɛks-spaɪ). I find the latter is slightly easier to pronounce.

> I like the exspy maybe the best because it also gives the idea that it used to be part of hyperspy.

Yes, but this can also be misleading/distracting.

jlaehne commented:
Thumbs up for *EleXSpy*

For me the *Spy* already contains spectroscopy and python, so anything with spectro would be a redundancy.

francisco-dlp commented:
I like the minimalist ex*Spy*: it contains all the relevant information (and maybe even too much) in the minimum number of characters.

ericpre commented:
We are close to be able to come back to splitting but it seems that we still two contenders for the package name: `exSpy` or `EleXSpy`, should we put it to a vote? 

| name             |  ballot paper  |
|---------------|-----------------|
| `exSpy`          |  :rocket:          |
| `eleXSpy`       |  :tada:            |

CSSFrancis committed:
Refactor: Fix failing tests after rebase

CSSFrancis committed:
Refactor: Change tests to remove installing holospy

CSSFrancis committed:
Refactor: Move EdgesRange

CSSFrancis committed:
Refactor: Move extra tests to spectrospy for plotting and extra

CSSFrancis committed:
Refactor: Moved material.py to exspy

CSSFrancis committed:
Refactor: Removed reference to exspy Signals in documentation

CSSFrancis committed:
BugFix: Fix readthedocs.yaml to install spectrospy

CSSFrancis committed:
Documentation: Cleaning up references to eds and eels in the documentation.

CSSFrancis committed:
Testing: Added changes to gui packages for testing.

CSSFrancis commented:
@ericpre I think I still need to adjust the minimum test to run with the gui packages.

Other than that I think most of the PR is getting closer to being done. There is a bit more cleaning up to do especially with the documentation.  I think a lot of it might stay for now as I think they are still good examples and the hyperspy documentation isn't really complete without them.

We still need to change the name of the package as well but that should be rather minimal. 

I don't if I'll be able to get back to this until Thursday morning so if anyone else wants to take this up feel free otherwise I will try to finish this then.

CSSFrancis committed:
Refactor spectrospy--> exspy

CSSFrancis committed:
Refactor: Change back links to gui tools

CSSFrancis committed:
Refactor: Add skipping tests if exspy isn't installed.

CSSFrancis committed:
Refactor: Add exspy to min versions

CSSFrancis committed:
Refactor: Move and remove eelsdb from datasets and Remove api.material from index.rst

CSSFrancis commented:
@ericpre do you want to look at this when you get the chance?  

I think that it is good to go.  At least most of the functionality has been split.  There are a couple of things that I don't know if I should move over such as the hard coded elements `hyperspy.misc.elements` and `hyperspy.misc.physics_tools` 

Part of me thinks that these parts of the code might be better in the eels package but I'm not sure.  At this point hyperspy 
is just Visualization, ML and some parallel computing. I'm pretty happy with that current state even if I think that the ML section needs a little triage ;)

CSSFrancis commented:
Just a fun little comment but after splitting the io the eels/eds and holography:
```
+----------+-----------------+-----------------+----------------------+
| Package  |  Lines of Code  |  Lines of Tests |   Lines of Comments  |
+----------+-----------------+-----------------+----------------------+
| Hyperspy |       33,889    |      22,810     |       22,805         |
+----------+-----------------+-----------------+----------------------+
| exspy    |        5,721    |       2,771     |        4,282         |
+----------+-----------------+-----------------+----------------------+
| rsciio   |       17,453    |      15,971     |        6,815         |
+----------+-----------------+-----------------+----------------------+
| pyxem    |       10,897    |      13,356     |       11,788         |
+----------+-----------------+-----------------+----------------------+
| holospy  |          603    |         371     |          519         |
+----------+-----------------+-----------------+----------------------+
```

This means that hyperspy is about half the size from when it started which should make changes to the code base much more manageable. These numbers are definitely effected by `black` but I still think this is a great change and really good for the health of the community. 

I realize that lines of code are a bad way to measure things but, I think that at this point there is a pretty good argument that hyperspy is in fact __not bloated__ as has been suggested to me in the past. ;) 

The release of hyperspy 2.0.0 should also remove a fair bit of code from pyxem meaning that it will likely shrink with time 😄

ericpre commented:
`hyperspy.misc.elements` needs to go in exspy package, it shouldn't stay in hyperspy. `hyperspy.misc.physics_tools` is not used anywhere and should be removed in a separate PR, it is trivial code that doesn't have its place in hyperspy (or exspy).

CSSFrancis commented:
> `hyperspy.misc.elements` needs to go in exspy package, it shouldn't stay in hyperspy. `hyperspy.misc.physics_tools` is not used anywhere and should be removed in a separate PR, it is trivial code that doesn't have its place in hyperspy (or exspy).

Okay I'll pull out the elements and leave the physics tools for a separate PR.  I need to clean up my commit history a bit and fix the failing tests.  I was a little too confident before pushing some of the changes 😅

ericpre commented:
Where does the `exspy.gui`  module come from? It is something that comes from the git history when splitting?

ericpre commented on exspy/exspy/misc/eds/eds.py:
```diff
@@ -1,22 +1,23 @@
 
-from hyperspy.misc.eds.utils import (
+
+from exspy.exspy.misc.eds.utils import (
```
```suggestion
from exspy.misc.eds.utils import (
```
?

ericpre reviewed (commented):
Should the `EDS_SEM_Spectrum` and `EDS_TEM_Spectrum` dataset be moved too? It is a bit of a pain because it is used in example, but it does need to go, I think!

CSSFrancis commented:
> Where does the `exspy.gui` module come from? It is something that comes from the git history when splitting?

This comes from the `hyperspy_gui_traitsui` and `hyperspy_gui_ipywidgets` packages.  Maybe this causes problems related  running with no_gui but I'm not sure.  Otherwise a separate `gui`  package for exspy with only a little code seems silly. 

Maybe it's just complicated as we are splitting things but I was getting frustrated with juggling 6 different repos and getting the tests to pass. 

> My current plan is that we should be able to `git-filter-repo` the commits from `hyperspy`,` hyperspy_gui_traitsui` and `hyperspy_gui_ipywidgets` so that all of the EELS EDS stuff is in one place.

This was my idea for how to get things to work?

CSSFrancis commented:
> Should the `EDS_SEM_Spectrum` and `EDS_TEM_Spectrum` dataset be moved too? It is a bit of a pain because it is used in example, but it does need to go, I think!

Yes they should be moved...  We need to think of the best way to not have our test coverage drop.  We can move them and still depend on them for testing but __ideally__ the tests wouldn't depend on that.

CSSFrancis committed:
Refactor: moved hyperspy.misc.elements to exspy

CSSFrancis commented:
@ericpre Okay sorry for the confusion :) Everything __should be__ more clean now (and hopefully the tests actually pass)

ericpre commented:
> Yes they should be moved... We need to think of the best way to not have our test coverage drop. We can move them and still depend on them for testing but ideally the tests wouldn't depend on that.

The test needs to be updated to use generic signal. It doesn't seem to be used in many places, so not too much of a pain!

ericpre commented on exspy/doc/intro.rst:
```diff
@@ -0,0 +1,104 @@
+
+What is Spectrospy
```
Do you want to rename `Spectrospy` to `exSpy` now or leave to when the repository is setup?

ericpre commented:
> > Where does the `exspy.gui` module come from? It is something that comes from the git history when splitting?
> 
> This comes from the `hyperspy_gui_traitsui` and `hyperspy_gui_ipywidgets` packages. Maybe this causes problems related running with no_gui but I'm not sure. Otherwise a separate `gui` package for exspy with only a little code seems silly.
> 
> Maybe it's just complicated as we are splitting things but I was getting frustrated with juggling 6 different repos and getting the tests to pass.
> 
> > My current plan is that we should be able to `git-filter-repo` the commits from `hyperspy`,` hyperspy_gui_traitsui` and `hyperspy_gui_ipywidgets` so that all of the EELS EDS stuff is in one place.
> 
> This was my idea for how to get things to work?

I was expecting to leave the gui code in their respective packages and not touch it? But yes, both approach are possible! I have a slight preference to keep it as it is because it means that the `exSpy` doesn't have to deal with gui dependencies at all and for maintenance (ci, running test suite) and installation, it is easier.

ericpre commented on hyperspy/tests/signals/test_assign_subclass.py:
```diff
@@ -141,27 +136,31 @@ def setup_method(self, method):
         self.s = hs.signals.Signal1D([0, 1])
 
     def test_lazy_to_eels_and_back(self):
+        exspy = pytest.importorskip("exspy")
```
I think these tests are redundant, in hyperspy, we only need one test to check that setting a `signal_type` change the object class but that's all.

Testing assignment to signal specify class should go in `exSpy`

ericpre reviewed (commented):
We need to move the preferences too:

https://github.com/hyperspy/hyperspy/blob/70ca0762a332a7239a4c1eb75001062ab6c7495d/hyperspy/defaults_parser.py#L119-L123

https://github.com/hyperspy/hyperspy/blob/70ca0762a332a7239a4c1eb75001062ab6c7495d/hyperspy/defaults_parser.py#L180-L199

CSSFrancis commented:
> I was expecting to leave the gui code in their respective packages and not touch it? But yes, both approach are possible! I have a slight preference to keep it as it is because it means that the `exSpy` doesn't have to deal with gui dependencies at all and for maintenance (ci, running test suite) and installation, it is easier.

We can't leave it in their respective packages or else the gui for just hyperspy doesn't work without exspy installed? Or do I have that wrong and the extensions are happy when  there are functions without toolkits? 

We can always have the gui stuff in exspy and then have it be an optional install? Correct? or does that make all of the tests harder to run.  I'll admit the gui toolkit stuff has been kind of a pain to split so I'm open to all suggestions. 

> The test needs to be updated to use generic signal. It doesn't seem to be used in many places, so not too much of a pain!

Can you do that possibly?  I don't have a great understanding of the peak fitting stuff in `hyperspy` so writing new tests takes me a bit longer than it should :)

ericpre commented:
> We can't leave it in their respective packages or else the gui for just hyperspy doesn't work without exspy installed? Or do I have that wrong and the extensions are happy when there are functions without toolkits?

I haven't check but I would expect that it would work! Using `hyperspy_gui_*` as it is without having `expy` installed should work through the registry: each library declare their class/function in the `ymal` file, these get registered and these are when necessary.

> Can you do that possibly? I don't have a great understanding of the peak fitting stuff in hyperspy so writing new tests takes me a bit longer than it should :)

Yes, I can do this.

CSSFrancis commented:
> I haven't check but I would expect that it would work! Using `hyperspy_gui_*` as it is without having `expy` installed should work through the registry: each library declare their class/function in the `ymal` file, these get registered and these are when necessary.

@ericpre Hmm okay, I was kind of assuming that was the case when I started but for some reason it wasn't working.  That's probably just because something got moved and the whole `.ymal`  extensions got messed up.  Getting everything to talk nicely with lots of different version etc has been not the easiest thing which is why I changed things. 

I might not get around to giving this another shot until the weekend. If you want to take a crack at it feel free.

ericpre commented:
Okay, I will give a go.

ericpre committed:
Remove GUI code

ericpre committed:
Move EELS and EDS preferences to `exSpy`

ericpre committed:
Run exSpy test suite on CI

jlaehne commented on exspy/exspy/components/eels_arctan.py:
```diff
@@ -1,20 +1,21 @@
 # -*- coding: utf-8 -*-
-# Copyright 2007-2023 The HyperSpy developers
+# Copyright 2007-2023 The exspy developers
```
In some file headers, `exSpy` is capitalized correctly, but in most it is still all lower case.

jlaehne commented on exspy/exspy/release_info.py:
```diff
@@ -0,0 +1,6 @@
+name = "exspy"
+version = "0.0.1"
+author = "exspy Developers"
```
```suggestion
author = "exSpy Developers"
```

jlaehne commented on exspy/exspy/release_info.py:
```diff
@@ -0,0 +1,6 @@
+name = "exspy"
+version = "0.0.1"
+author = "exspy Developers"
+copyright = "Copyright 2016-2023, The exspy developers"
```
```suggestion
copyright = "Copyright 2016-2023, The exSpy developers"
```

jlaehne replied:
Why from 2016? It is based on eelsLab, which is much older.

jlaehne commented on exspy/exspy/signals/__init__.py:
```diff
@@ -0,0 +1,33 @@
+
+""" Signals to be operated on. The basic unit of data"
+    EELSSpectrum
+        For electron energy-loss data with signal_dimension equal 1, i.e.
+        spectral data of n-dimensions. The signal is binned by default.
```
```suggestion
        For electron energy-loss data with signal_dimension equal one, i.e.
        spectral data of `n` navigation dimensions. The signal is binned by default.
```

jlaehne commented on exspy/exspy/signals/__init__.py:
```diff
@@ -0,0 +1,33 @@
+        For electron energy-loss data with signal_dimension equal 1, i.e.
+        spectral data of n-dimensions. The signal is binned by default.
+    EDSTEMSpectrum
+        For electron energy-dispersive X-rays data acquired in a transmission
+        electron microscopy with signal_dimension equal 1, i.e.
+        spectral data of n-dimensions. The signal is binned by default.
```
```suggestion
        For electron energy-dispersive X-ray data acquired in a transmission
        electron microscope with signal_dimension equal one, i.e.
        spectral data of `n` navigation dimensions. The signal is binned by default.
```

jlaehne commented on exspy/exspy/signals/__init__.py:
```diff
@@ -0,0 +1,33 @@
+        electron microscopy with signal_dimension equal 1, i.e.
+        spectral data of n-dimensions. The signal is binned by default.
+    EDSSEMSpectrum
+        For electron energy-dispersive X-rays data acquired in a scanning
+        electron microscopy with signal_dimension equal 1, i.e.
+        spectral data of n-dimensions. The signal is binned by default.
```
```suggestion
        For electron energy-dispersive X-ray data acquired in a scanning
        electron microscop with signal_dimension equal one, i.e.
        spectral data of `n` navigation dimensions. The signal is binned by default.
```

jlaehne commented on exspy/exspy/signals/__init__.py:
```diff
@@ -0,0 +1,33 @@
+        electron microscopy with signal_dimension equal 1, i.e.
+        spectral data of n-dimensions. The signal is binned by default.
+    DielectricFunction
+        For dielectric function data with signal_dimension equal 1. The signal
+        is unbinned by default.
```
```suggestion
        For dielectric function data with signal_dimension equal one. The signal
        is unbinned by default.
```

ericpre committed:
Rename `Model1D.low_loss` to `Model1D.convolve_signal`

ericpre committed:
Improve documentation

ericpre committed:
Move EELS/EDS datasets to `exSpy`

ericpre committed:
Use `exSpy` consistently

ericpre committed:
Improve documentation

ericpre committed:
Move `VolumePlasmaDrude` to `exSpy`

jlaehne commented:
References in HyperSpy's `CHANGES.rst` to methods that are now in exSpy will have to use intersphinx.

ericpre commented:
Yes, for the changelog, it will need updating once the exspy's documentation is live. There are some follow up tweaks needed for the exspy's documentation, but it is not convenient to work on this in this PR. What's matter here is to transfer the relevant bits of code between hyperspy and exspy and as far as I can tell, this should be all done!

ericpre commented:
Actually some tests could be moved from hyperspy to exspy.

ericpre committed:
Tidy up CI

ericpre committed:
Move tests to `exSpy`

francisco-dlp commented:
Awesome job!

I think that there are a few more things to do here:

- [ ] Move the convolution feature to exSpy: this sort of convolution only makes sense for CL EELS
- [x] Move the following photoelectron spectroscopy compoents to exSpy:
    - [x] `pes_core_line_shape`
    - [x] `pes_see`
    - [x] `pes_voigt`
 - [ ] Audit the code to identify any API changes worth doing at this point such as privatizing attributes (e.g., `channel_switches`) and methods.

I can work on this next week, as long as somebody is not already working on any of these points.

ericpre committed:
Move EELS/EDS part of metadata structure to `exspy`

ericpre commented:
> * [ ]  Move the convolution feature to exSpy: this sort of convolution only makes sense for CL EELS

Could it be useful in other scenario, for example, to convolve with a point source? The convolution feature itself is generic and not signal specific - even, its know usage is for `EELSModel`.

> * [ ]  Audit the code to identify any API changes worth doing at this point such as privatizing attributes (e.g., `channel_switches`) and methods.

Do you have something else in mind?

ericpre committed:
Improve documentation

ericpre committed:
Move PES components to `exspy`

francisco-dlp commented:
Thans for moving the PES components.

While convolving the model could be useful beyond EELS, the way in which the feature is implemented is EELS specific. In particular, the model is evaluated on an axis customized for EELS core-loss (extended to the left to avoid border artifacts). One possibility could be moving only the method that generates the convolution axis to exSpy or (better) generalizing that method to make it useful beyond EELS, and calling it with the right parameters in exSpy.

ericpre committed:
Privatise `channel_switches`

ericpre commented:
> One possibility could be moving only the method that generates the convolution axis to exSpy or (better) generalizing that method to make it useful for beyond EELS, and calling it with the right parameters in exSpy.

Yes, it makes sense, and should fairly simple to update/move `set_convolution_axis`? https://github.com/hyperspy/hyperspy/blob/a3a038194d9168b2535c9f1b4003a6b953b93c5a/hyperspy/models/model1d.py#L332-L343C50

I did the `channel_switches` bits, is there anything else that will be needed in this PR?

jlaehne commented:
> While convolving the model could be useful beyond EELS, but the way in which the feature is implemented is EELS specific.

Indeed, in LumiSpy, we are interested in creating components for luminescence transients that involve a convolution. So it might be useful if that functionality could generalized.

francisco-dlp commented:
@jlaehne, can you give more details on the convolution feature that you need in LumiSpy?

The current feature convolves a `Signal` with the model of the `Signal` that we are fitting. Both signals must have the same navigation dimensions. Furthermore, the model is evaluated on an axis that is extended to the left, then convolved and cropped to the original axis. This is all very core-loss EELS specific, so I still think that it might be better to move the feature to exSpy.

For general usage, I could imagine convolving the whole model with a component (typically a peaky one, such as a Gaussian). The parameters in this case could be free (something not possible with the current implementation). However, nobody has yet requested such  a feature...

jlaehne commented:
> can you give more details on the convolution feature that you need in LumiSpy?

In LumiSpy, we would want to convolute a system response function with a decay model.

The system response function could be
- a Gaussian or
- a hyperbolic secant component,
- but possibly also a measured curve (so a `signal` as in the EELS case). The decay model could 

The decay model could be e.g.
- one or multiple exponentials
- a stretched exponential
- ...

Usually, the rise time of the system response is included in the experimental data, so we probably would not have to extend the model as in EELS, but I might have a closer look at your implementation.

francisco-dlp commented:
Thanks for the details @jlaehne. I have some questions:

- Is the decay model part of a larger model?
- When the system response is a signal, is it different at each pixel?

jlaehne commented:
>     * Is the decay model part of a larger model?
As indicated above, the model can have 1-3 components for the decay (and that depends on the sample, but would usually be the same across a map) and this is convolved with the system response. The parameters of these components change of course from pixel to pixel. So far, I do not see any overarching larger model.

>     * When the system response is a signal, is it different at each pixel?
No, the system response function should be the same for every pixel in the navigation dimension.

francisco-dlp commented:
In that case, I think that the convolution feature implemented for EELS won't be of much use for your application.

ericpre commented:
@francisco-dlp, do you know if you would have today or tomorrow to do the convolution bits? If not, I will try to do that later today in the interest of finishing this PR.
As I understand, the approach currently used for setting the convolution axis is valid but suboptimal (1. larger axis due to extrapolation makes slower fit, 2. this is not implementation for linear fitting without iterating over the navigation space) in more simple convolution case, where no extrapolation is necessary. If my understand if correct, we simply need to move the `set_convolution_axis` to `EELSModel` and make a more simple version in `Model1D`, using the `Model1D.axis.axis` without extrapolation?
The improvement with linear fitting can be done in a separate PR, as this is unrelated to the exSpy split.

francisco-dlp commented:
I can start working on it tomorrow.

I still think that, as it is, the feature is too EELS-specific. Therefore, I propose to move it all to exSpy. If somebody needs convolving the whole model in the future, chances are that we'll need to take a different approach anyway since most usage cases involve convolving with a single signal that could possible be a component with free parameters. Does it make sense?

ericpre commented:
Aside the axis/model extrapolation, what else is left that is EELS specific? With moving it all to exSpy, do you mean removing the convolution functionality from `Model1D`? We just need to add supports for single signal, so that it can be more straightforward to use with more simple case than EELS. Then it will be easy to implement the convolution for linear fit without iteration over the navigation.

francisco-dlp commented:
Yes, I mean removing the convolution functionality from `Model1D`. This is because, if we simply more `set_convolution_axis`, the rest of the code is not usable as is. Therefore, I propose to move all to exSpy. This has the advantage that we can rename `convolve_signal` back to `low_loss` to enhance backward compatibility. If HyperSpy ever needs a convolution function based on the existing one we can always move it back.

CSSFrancis commented:
@ericpre and @francisco-dlp what is left that needs to be done here?  I'm for moving the convolution stuff to `exspy` if that is the thing holding things up.

francisco-dlp commented:
@CSSFrancis, I was planning to do it myself. I have already started, but I won't be able to come back to it until Thursday afternoon. If you have already started yourself and you feel like it, go on, but do let me know.

CSSFrancis commented:
> @CSSFrancis, I was planning to do it myself. I have already started, but I won't be able to come back to it until Thursday afternoon. If you have already started yourself and you feel like it, go on, but do let me know.

I haven't started anything yet :) It would probably would take me a little bit to familiarize myself with the code, just let me know if you need me to look at anything or review any additions.

francisco-dlp committed:
Move convolution Model.__call__ functionality to exSpy

francisco-dlp committed:
Move Model._jacobian convolution code to exSpy

ericpre commented:
@francisco-dlp, do you known when you will be able to move the convolution to exspy? Looking at what needs to be done, it seems not straightforward as it is used in the calculation of the jacobian, and other places. If it can't done in a reasonable timeline, I would suggest to leave it for later, because this is holding the progress toward releasing hyperspy 2.0 - what is left should be some tidying up to be done after exspy is split.

francisco-dlp commented:
Yes, it is not straightforward. I am finishing it. I will submit a PR to this branch today.

francisco-dlp committed:
Move linear fitting convolution code to exSpy

francisco-dlp committed:
Document .

francisco-dlp committed:
Add  `Model.convolved` setter for backward compatibility

francisco-dlp committed:
Move plot model convolution tests to exSpy

francisco-dlp committed:
Fix exspy model plot test and add updated test figures

francisco-dlp committed:
Move EELS file test to exspy

francisco-dlp committed:
Rename  back to

francisco-dlp committed:
Fix test

francisco-dlp commented:
Here is the PR: https://github.com/CSSFrancis/hyperspy/pull/2

Moving the code required some small structural changes that should make it easier to implement convolution in other `BaseModel` subclasses.

Some convolution code remains in HyperSpy:

- The `Component.convolved` attribute
- The `Model.convolved` attribute for back-compatibility
- Linear fit code for convolution in `Expression` (there is no way to move it elsewhere)
- A few harmless lines of code in SAMFIRE check the `Model.convolved` attribute among others. Since that part of the code is not covered by tests, and SAMFIRE requires some care anyway, I prefer no to attempt to move it at this stage.

francisco-dlp committed:
Move IO test back to where it belongs, thanks @ericpre

CSSFrancis committed:
Merge pull request #2 from francisco-dlp/MV_convolution

Move convolution features to exSpy

CSSFrancis committed:
BugFix: Fix backwards compatibility with missing exspy package and older Components

CSSFrancis committed:
Testing: Change filereader_metadata to account for moved files

CSSFrancis commented:
@francisco-dlp I (think) I fixed the failing tests.  One was pointing to a file which didn't exist.  Loading some of the older Models was also failing if you didn't have `exspy` installed and wasn't pointing towards installing `exspy` so I fixed that.

CSSFrancis commented:
@ericpre The failure is related to the pooch/zendo bug correct?

https://github.com/fatiando/pooch/issues/371 We should be able to run the CI again once pooch 1.8.0 hits conda in an hour or two.

CSSFrancis commented:
Thoughts on merging/splitting and then figuring stuff like the documentation out separately?

ericpre committed:
Fix packaging tests exspy

ericpre reviewed (approved):
Azure pipelines CI is still failing because pooch 1.8 is still not on conda-forge, it should be a couple of hours.

@francisco-dlp, do you want to do another review?

CSSFrancis commented:
@ericpre  @francisco-dlp Do we want to merge this and then we can try to clean some stuff in later PR's if needed.  I think this PR is mostly good (at least 3 people have looked at the changes and the CI passes.)

ericpre commented:
I would suggest to leave it up open for comments to tomorrow, and if this is none, then we merge?

francisco-dlp committed:
Revert "Testing: Change filereader_metadata to account for moved files"

This reverts commit c5088e3aef340b02fafdc3aef4613f9f261fc034.

francisco-dlp committed:
Delete wrong file path

francisco-dlp committed:
Add test file back

francisco-dlp committed:
Add back pytest decorator removed in previous revert

francisco-dlp commented on hyperspy/model.py:
```diff
@@ -126,11 +126,18 @@ def reconstruct_component(comp_dictionary, **init_args):
+                'failed because the component is provided by the ' +
+                f'{comp_dictionary["package"]} Python package, but ' +
+                f'{comp_dictionary["package"]} is not installed.')
+        except KeyError:
```
Would it be possible to make this more precise e.g. by checking if the component's package is HyperSpy and `comp_dictionary["class"]` is one of:

```python
EXSPY_HSPY_COMPONENTS = ("EELSArctan", "EELSCLEdge", "DoublePowerLaw", "Vignetting", "PESCoreLineShape", "SEE", "PESVoigt", "VolumePlasmonDrude")
```

?

francisco-dlp commented:
I have sent a PR (https://github.com/CSSFrancis/hyperspy/pull/3) to fix the path of the file that @CSSFrancis though that we have moved to exSpy.

Other than this and the small comment above, this LGTM!

CSSFrancis committed:
Merge pull request #3 from francisco-dlp/FIX_moved_test_file

Fix moved test file

CSSFrancis committed:
Refactor: Account for moved exspy components

CSSFrancis commented:
> I have sent a PR ([CSSFrancis#3](https://github.com/CSSFrancis/hyperspy/pull/3)) to fix the path of the file that @CSSFrancis though that we have moved to exSpy.
> 
> Other than this and the small comment above, this LGTM!

Done!

CSSFrancis commented:
LGTM?

CSSFrancis commented:
@ericpre Do you want to merge this and do the git filter?  I can give it another shot otherwise but I'm not sure how great I did last time 😅

ericpre commented:
I would like to check briefly the last commit and #3238 also needs to be merge before splitting.

I am happy to do it in the coming days.

ericpre committed:
Add test for loading component when exspy is not installed

ericpre commented:
For reference, this is the command that I have used to split to https://github.com/hyperspy/exspy:
```
git filter-repo --path exspy --path spectrospy --path hyperspy/tests/utils --path hyperspy/tests/misc --path hyperspy/tests/models --path hyperspy/tests/components --path hyperspy/tests/drawing/plot_model --path hyperspy/tests/drawing/plot_spectra_markers --path hyperspy/tests/drawing/data --path hyperspy/tests/drawing/test_plot_spectra_markers.py --path hyperspy/tests/drawing/test_plot_model.py --path hyperspy/tests/signals --path hyperspy/signal_tools.py --path hyperspy/docstrings/model.py --path hyperspy/misc/elements.py --path hyperspy/misc/material.py --path hyperspy/misc/eds  --path hyperspy/misc/eels --path examples/model_fitting --path hyperspy/_signals --path hyperspy/signals --path hyperspy/_components --path hyperspy/components --path eelslab --path lib --path src --path-rename exspy/: --force
```
It is quite long to cherry pick the relevant folders from the history and remove the files that are not relevant (and sometimes large, e.g. io test file, etc.).
